### PR TITLE
fix reported bugs

### PR DESCRIPTION
**Pull request by rexagod**
fixes https://github.com/publiclab/Leaflet.DistortableImage/issues/306

test_

cc @jywarren @sashadev-sky 

Fixes #0000 (<=== Add issue number here)

Make sure these boxes are checked before your pull request (PR) is ready to be reviewed and merged. Thanks!

* [x] PR is descriptively titled 📑 and links the original issue above 🔗
* [x] tests pass -- look for a green checkbox ✔️ a few minutes after opening your PR -- or run tests locally with `grunt`
* [x] code is in uniquely-named feature branch and has no merge conflicts 📁
* [x] screenshots/GIFs are attached 📎 in case of UI updates
* [x] @mention the original creator of the issue in a comment below for help or for a review

> We're happy to help you get this ready -- don't be afraid to ask for help, and **don't be discouraged** if your tests fail at first!

If tests do fail, click on the red `X` to learn why by reading the logs.

Please be sure you've reviewed our contribution guidelines at https://publiclab.org/contributing-to-public-lab-software 

Thanks!

### CODE REVIEW

**Comment by rexagod** [comment]
- Should I decrease the disappearance time of the tooltips and icons to 1 sec (from 2 sec), or is it okay?
- "Mo found that the button that seemed to make sense was "rotate" but that didn't override the lock. Perhaps we should ensure that clicking rotate/scale/distort unlocks the image automatically." I think this should be incorporated as well? From the discussion at #306? Just let me know and I'll get to it!

File: src/edit/LockHandle.js [hunk 1]
```diff
@@ -2,7 +2,7 @@ L.LockHandle = L.EditHandle.extend({
...
...JnNBrjxzJJpFWLSKFGI0YscVgiFZQehxkyYsSIMWNT9yq1W9XULryKDenxbbpSqZDn8TinzeVwZM6nUrtVPQFhYjdiyIghL/GUGg6FbDQ8SpchQxOVxnEDXF27VVMpikcDBrT5GqelUsHwOcwRBvFryEgRafUTu8CkBAP69DnOl7ik9K5QZ/EM36BDnz79WCNfNdJaRKQRQ/r06NHlab4slQp0/i7yF5wxGoUqRcldIc+hW9gTEUakATX69PCocYAhH2aL1qEoQFI3z9c4Qte8evQZMiyuRkUWKWp/D+hRw8XF5iku8X5uo46l0TqxZ27IKzzC63TiV5cegzgeFbJGcgt6MiwCfEbYOLi4ONhYWLzCF7iVd3MTDWzpNIESneVxnqVNjw7tWKR+SiRFpFVPEHxGDHCwjUbh3x3kBbZyM7ewhSYejvk3JNaqixOl4T4jFjnBsxw12nTp0GGRtolHQ0ZFfh7JKt77tqLfLBwcang0aDJFy/xqUKdOjRo1POo0aVKnTh3bxC6k1gpKE87MDyf8DBnRp2/St2gOQ5++EalNmzY9czcpNUGoaOPSLfSp840KVtx+CO8sRSIlsepycaTQyiqV/jpcJ2gcV7WJSF069OgzYlTkp5GKLFJUJ6XXoAkb4mmRwtrJjnVLYpFYeZ2C1LkJjEZDhvTp04v7ddn6SDXSGp2scSo+RXeW6tTw4iaEHUek9C+xOiIlazEkc1EilfoMTDQaFVujoqd2llEpe8Xr45mGeDoiWXE0spTcrUpql6gUPcscTU8dmilBg9Qcu4Iv81n0iGSZK15yZ2lIjX5Ko3SNZCkerUn7IVJpbOajjMx04/RE1ULHoyJ37bJ/tLCxsXFwcIxEDpZpfX+/xE5q5dlguFJyF8UkP5YpepQvaTEE3+8bSqSVFIlYDyulk21e1jKRLv/fJdHKyZRePTW9UpCfUugqsUgirYVI6Y6cZWqitERZhdQCXx2ZltdKPum1n4LMWnZIpMkQiUzEscDcdr28LpI8a9d4CFIVLVeXSCKtrUhXizdv9DdidSPTD7zCt2Y2TM4pDKTNxDckSoOlJ+GEuHH0EJwQEkkIiSSERBJCSCQhJJIQEkkIiSSEkEhCSCQhJJIQEkkIIZGEkEhCSCQhJJIQQiIJIZGEkEhCSCQhhEQSYk34/55dj6jwQIHlAAAAAElFTkSuQmCC',
```
Comment (rexagod) [inline src/edit/LockHandle.js]: >  It was hard to tell that the Xs mean "locked" -- seemed ambiguous or unexplained

I replaced 'X'es with 'Lock' images here for a more semantic UI.

File: assets/icons/svg/unlock.svg [hunk 2]
```diff
@@ -1,3 +1,3 @@
...
-  <path fill="#0078A8" fill-rule="nonzero" d="M13.5,6 L12.75,6 L12.75,4.5 C12.75,2.43 11.07,0.75 9,0.75 C6.93,0.75 5.25,2.43 5.25,4.5 L6.75,4.5 C6.75,3.255 7.755,2.25 9,2.25 C10.245,2.25 11.25,3.255 11.25,4.5 L11.25,6 L4.5,6 C3.675,6 3,6.675 3,7.5 L3,15 C3,15.825 3.675,16.5 4.5,16.5 L13.5,16.5 C14.325,16.5 15,15.825 15,15 L15,7.5 C15,6.675 14.325,6 13.5,6 Z M13.5,15 L4.5,15 L4.5,7.5 L13.5,7.5 L13.5,15 Z M9,12.75 C9.825,12.75 10.5,12.075 10.5,11.25 C10.5,10.425 9.825,9.75 9,9.75 C8.175,9.75 7.5,10.425 7.5,11.25 C7.5,12.075 8.175,12.75 9,12.75 Z"/>
+  <path fill="#000" fill-rule="nonzero" d="M13.5,6 L12.75,6 L12.75,4.5 C12.75,2.43 11.07,0.75 9,0.75 C6.93,0.75 5.25,2.43 5.25,4.5 L5.25,6 L4.5,6 C3.675,6 3,6.675 3,7.5 L3,15 C3,15.825 3.675,16.5 4.5,16.5 L13.5,16.5 C14.325,16.5 15,15.825 15,15 L15,7.5 C15,6.675 14.325,6 13.5,6 Z M6.75,4.5 C6.75,3.255 7.755,2.25 9,2.25 C10.245,2.25 11.25,3.255 11.25,4.5 L11.25,6 L6.75,6 L6.75,4.5 Z M13.5,15 L4.5,15 L4.5,7.5 L13.5,7.5 L13.5,15 Z M9,12.75 C9.825,12.75 10.5,12.075 10.5,11.25 C10.5,10.425 9.825,9.75 9,9.75 C8.175,9.75 7.5,10.425 7.5,11.25 C7.5,12.075 8.175,12.75 9,12.75 Z"/>
```
Comment (rexagod) [inline assets/icons/svg/unlock.svg]: >  could we show the lock button depressed while it's locked?

Changed the path fill color to black to show its "active" state.

File: src/edit/LockHandle.js [hunk 3]
```diff
@@ -12,6 +12,59 @@ L.LockHandle = L.EditHandle.extend({
...
+        iconAnchor: [ 25, 50 ]
+      }
+    });
+   
+	  var question = new Icon({
+      iconUrl: '../assets/icons/markers/question-mark.png'
+    });
+	  
+	  this._questionMark = new L.marker(this._handled.getCorner(this._corner), { icon: question })
+		  .bindTooltip("Locked!")
```
Comment (rexagod) [inline src/edit/LockHandle.js]: could we show a tooltip over the X handles that says Locked or Image is locked?

Done here.

File: src/edit/DistortableImage.Edit.js [hunk 4]
```diff
@@ -355,14 +354,12 @@ L.DistortableImage.Edit = L.Handler.extend({
   },
 
   _toggleTransparency: function() {
-    var image = this._overlay._image,
-      opacity;
+    var image = this._overlay._image;
 
-    this._transparent = !this._transparent;
-    opacity = this._transparent ? this.options.opacity : 1;
+    this._opacity = Math.round(this._opacity*10) === 1 ? 1 : this._opacity - 0.3;
```
Comment (rexagod) [inline src/edit/DistortableImage.Edit.js]: > Get the transparency tool to work on a range

Done here.

File: src/edit/LockHandle.js (hunk 1 above)
Comment (jywarren) [reply src/edit/LockHandle.js]: Hmm, I have a preference for the Xs, but I love the tooltips. Would you mind changing them back? I just find they are least likely to cover up something, and don't create visual clutter. Thanks!

**Comment by jywarren** [comment]
This is great. Once the lock/X is changed i think this is good but want to hear from @sashadev-sky too! Thanks, @rexagod !!!

**Comment by jywarren** [comment]
Hm, and to me, the ? icons on tooltips are a bit cluttered as well. My preference would be for just tooltips! Thanks so much for this! 🎉

**Comment by IgorWilbert** [comment]
@rexagod turning the question mark balloon down or putting the tooltip below the square should solve the cluttering. Thanks for the PR!

File: dist/leaflet.distortableimage.js [hunk 5]
```diff
@@ -1129,7 +1129,7 @@ L.LockHandle = L.EditHandle.extend({
...
...JnNBrjxzJJpFWLSKFGI0YscVgiFZQehxkyYsSIMWNT9yq1W9XULryKDenxbbpSqZDn8TinzeVwZM6nUrtVPQFhYjdiyIghL/GUGg6FbDQ8SpchQxOVxnEDXF27VVMpikcDBrT5GqelUsHwOcwRBvFryEgRafUTu8CkBAP69DnOl7ik9K5QZ/EM36BDnz79WCNfNdJaRKQRQ/r06NHlab4slQp0/i7yF5wxGoUqRcldIc+hW9gTEUakATX69PCocYAhH2aL1qEoQFI3z9c4Qte8evQZMiyuRkUWKWp/D+hRw8XF5iku8X5uo46l0TqxZ27IKzzC63TiV5cegzgeFbJGcgt6MiwCfEbYOLi4ONhYWLzCF7iVd3MTDWzpNIESneVxnqVNjw7tWKR+SiRFpFVPEHxGDHCwjUbh3x3kBbZyM7ewhSYejvk3JNaqixOl4T4jFjnBsxw12nTp0GGRtolHQ0ZFfh7JKt77tqLfLBwcang0aDJFy/xqUKdOjRo1POo0aVKnTh3bxC6k1gpKE87MDyf8DBnRp2/St2gOQ5++EalNmzY9czcpNUGoaOPSLfSp840KVtx+CO8sRSIlsepycaTQyiqV/jpcJ2gcV7WJSF069OgzYlTkp5GKLFJUJ6XXoAkb4mmRwtrJjnVLYpFYeZ2C1LkJjEZDhvTp04v7ddn6SDXSGp2scSo+RXeW6tTw4iaEHUek9C+xOiIlazEkc1EilfoMTDQaFVujoqd2llEpe8Xr45mGeDoiWXE0spTcrUpql6gUPcscTU8dmilBg9Qcu4Iv81n0iGSZK15yZ2lIjX5Ko3SNZCkerUn7IVJpbOajjMx04/RE1ULHoyJ37bJ/tLCxsXFwcIxEDpZpfX+/xE5q5dlguFJyF8UkP5YpepQvaTEE3+8bSqSVFIlYDyulk21e1jKRLv/fJdHKyZRePTW9UpCfUugqsUgirYVI6Y6cZWqitERZhdQCXx2ZltdKPum1n4LMWnZIpMkQiUzEscDcdr28LpI8a9d4CFIVLVeXSCKtrUhXizdv9DdidSPTD7zCt2Y2TM4pDKTNxDckSoOlJ+GEuHH0EJwQEkkIiSSERBJCSCQhJJIQEkkIiSSEkEhCSCQhJJIQEkkIIZGEkEhCSCQhJJIQQiIJIZGEkEhCSCQhhEQSYk34/55dj6jwQIHlAAAAAElFTkSuQmCC',
```
Comment (IgorWilbert) [inline dist/leaflet.distortableimage.js]: That is one huge URL! Should we put it in another file ar a variable elsewhere?

File: dist/leaflet.distortableimage.js [hunk 6]
```diff
@@ -1139,6 +1139,59 @@ L.LockHandle = L.EditHandle.extend({
...
+	  
+	  setTimeout(function () {
+		  var el = [];
+		  var marks = Array.from(document.querySelectorAll(".leaflet-interactive"));
+		  for (var i = 0; i < marks.length; i++) { 
+			  if (/question/.test(marks[ i ].currentSrc)) { 
+				  el.push(marks[ i ]);
+			  }
+		  }
+		  var tt = document.querySelector(".leaflet-tooltip");
```
Comment (IgorWilbert) [inline dist/leaflet.distortableimage.js]: Be careful with single and double letter variables ;)

File: src/edit/LockHandle.js [hunk 7]
```diff
@@ -12,6 +12,59 @@ L.LockHandle = L.EditHandle.extend({
 	_onHandleDrag: function() {
 	},
 
+  _bindListeners: function() {
+    this.on(
+      {
+        dragstart: this._onHandleDragStart,
```
Comment (IgorWilbert) [inline src/edit/LockHandle.js]: A snakecase `dragStart` would look great here

File: dist/leaflet.distortableimage.js (hunk 6 above)
Comment (rexagod) [reply dist/leaflet.distortableimage.js]: Sure! 😄 👍

File: src/edit/LockHandle.js (hunk 7 above)
Comment (rexagod) [reply src/edit/LockHandle.js]: It's actually a keyword (leaflet event) so its immutable.

File: dist/leaflet.distortableimage.js (hunk 5 above)
Comment (rexagod) [reply dist/leaflet.distortableimage.js]: Reverted back to Xs: https://github.com/publiclab/Leaflet.DistortableImage/pull/358#discussion_r305636537

File: src/edit/LockHandle.js (hunk 1 above)
Comment (rexagod) [reply src/edit/LockHandle.js]: Done!

**Comment by jywarren** [comment]
Hi, can we have an updated gif please? Thank you! 

Did you want to bump the version too? @sashadev-sky does this look good to you? 

Hopefully this and a version bump will make their way to MapKnitter, as we're seeing also some multiple image selection bugs that I think @sashadev-sky has fixed already... just need to bump the published version in MK... 👍

**Comment by rexagod** [comment]
@jywarren Here's the gif. I've also bumped the patch version.

test

**Comment by jywarren** [comment]
Ah, i think this may need a rebase and another version bump as Sasha just fixed an urgent issue and went to 0.6.5 already? Thank you!

**Comment by rexagod** [comment]
@sashadev-sky Can you verify if all your changes are intact in the rebase? Thanks!

**Comment by rexagod** [comment]
Fixing the svgs now!

**Comment by sashadev-sky** [comment]
sorry all!! will check this out asap

**Comment by sashadev-sky** [comment]
@rexagod @jywarren have only looked at the gif so far but really awesome work with the lock icon!! I always thought the `open lock` `close lock` toggle was confusing but wasn't sure what to do about it

**Comment by sashadev-sky** [comment]
@rexagod @jywarren ok so ill rebase this before merge so don't worry about that - but question, 
did you consider any other interfaces for making the icon appear selected, like graying it out a bit / making it look depressed, and decide to go with making it black? I like it and how it matches the color of the handles, but we also should consider how would integrate in the context of the entire toolbar interface? Making it black could stand out as an inconsistency.

- Tooltip previews on each icon for locked is cool definitely on board with that

**Comment by rexagod** [comment]
Sure the svg could use some mods, although I'm not well versed with them, but I'd definitely like to take a look over the source of these svgs that we are using right now and make some changes concurring to your suggestion above!

**Comment by sashadev-sky** [comment]
@rexagod so the wiki explains very easily how system works and where they come from - https://github.com/publiclab/Leaflet.DistortableImage/wiki/SVG-Icon-System I'm totally new to svg as of this project but I love it. You can use any icon you want, and customize it in any way via paths. Next I am looking to learn how to animate svg, to make a loader icon from scratch

**Comment by jywarren** [comment]
Just circling back to note that I worry about how many "Locked" tooltips would show with the delay -- maybe best just immediately hide it when you move mouse away? Take for instance this map: 

image

**Comment by sashadev-sky** [comment]
Maybe we could add a “freeze frame” tool via imgGroup.editing.disable. It’s disables image editing, so basically locked, but without the lock symbols 👍🏻👍🏻 it can be like the “picture perfect” tool

Sasha Boginsky
Goizueta Business School
Emory University | BBA Class of 2017
646.241.5756 | sasha.boginsky@emory.edu

**Comment by rexagod** [comment]
Okay I understand Jeff's suggestion and it seems easily implementable to me, however @sashadev-sky would it not be better to keep the tooltips here?

**Comment by sashadev-sky** [comment]
@rexagod @jywarren ah this has been such a tough one to think of a solution for. Every angle I take seems incorrect, the lock interface is even unclear to us as we develop it, struggling to decide if locked images should be movable or not, editable, exportable. I think that honestly - we should just scrap it. 

I have thought a lot about it and here is what really makes sense to me:

**Point 1, 2 and 3** address the logistics of locking in the classic sense - who is allowed to edit your content on a website

**Point 4, 5, and 6** address locking for the editing experience - something that I think we misinterpreted and should name "freezing"

1) adding a publish button to mapknitter, and all images are locked once the map is published. Locked doesn't have an interface, no markers, no toolbars.

2) If map is anonymous, then it will just always stay locked (`map.readOnly = true`). When we look at content other people have made on websites, we already know that we aren't going to be able to just go in and edit it! This is not an assumption people have.

3) If the user made it with an account, then only for them when they open the map can they edit it (`map.readOnly = false`). perhaps we can provide them with some sort of "edit" / "save" button to toggle. Additionally, they can be "admins" and invite other people who are allowed to toggle the "edit" / "save" mode when they are logged in to preserve synchronous editing.

4) As for the editing experience, we have moved in the direction that `locked` images cannot be moved, but can be edited (transparency, for ex.) and exported. Locked isn't really the word I would use to describe this. This would be like `frozen`. PS lets you freeze images by just painting red over them with a freeze tool. 

5) the user can really get by without having to freeze out images given that our collection tools already kind of work in the _opposite way_ - they only work on images that are "painted yellow", the default isn't that this all one big canvas. Since our interface is already focused on the individual image, there is no need to have to lock it. 

6) if we really wanted, we could move forward with a 'freeze tool' later for cases where maybe the user wants to select a chunk of images at once with box select but make sure some aren't selected.

**Comment by sashadev-sky** [comment]
closing this as its stale / resolved

**Pull request closed without merging**